src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,122 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "antlr4-runtime.h"
+#include "src/common/logging/logging.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+
+namespace raksha::ir::auth_logic {
+
+static Principal ConstructPrincipal(
+    AuthLogicParser::PrincipalContext& context) {
+  return Principal(context.ID()->getText());
+}
+
+static datalog::Predicate ConstructPredicate(
+    AuthLogicParser::PredicateContext& context) {
+  std::vector<std::string> args = utils::MapIter<std::string>(context.pred_arg(), [](auto* statement) { return statement->ID()->getText(); });
```

Comment:
Nit: This looks longer than 80 characters. Please split at 80. Note that the clang formatting tool will do this kind of thing for you and Raksha is set up for it to format with Google style. I'm writing this comment on a tiny laptop screen, so disregard if already small enough.